### PR TITLE
Refactor `gradient_check.check_backward`

### PR DESCRIPTION
**Pull request by niboshi**
In the current implementation of  `gradient_check.check_backward`, computation of backward gradients / numerical gradients / some checks are mixed and very difficult to understand.

I refactored the function for readability and maintainability.

### CODE REVIEW

**Comment by niboshi** [comment]
I want to make it explicit that `_directional_backward_gradients` is used to update `y_grad`, so I made it return `y_grad` instead of updating.
Do you think its ok?

**Comment by niboshi** [comment]
Fixed so that the forward pass is done before direction sampling.
PTAL

File: chainer/gradient_check.py [hunk 1]
```diff
@@ -301,8 +301,220 @@ def _as_tuple(x):
...
+        y = _as_tuple(y)
+        y0_data = [_.data for _ in y]
+
+        # All creators of `y` need to be the same because we only call
+        # `y[0].backward` to call `backward` method of the creator.
+        # To do so we need to insert a dummy function `_GradientSetter` to the
+        # computational graph.
+        # Note that `func` may not be a `Function` object.
+
+        y, y_grad = _set_y_grad(y, y_grad)
```
[the remark this answers is not in the document]
Comment (toslunar) [reply chainer/gradient_check.py]: :+1: > https://github.com/chainer/chainer/pull/5411#issuecomment-429508346

**Comment by toslunar** [comment]
- This PR is intended to keep the semantics.
- All the difference I found is that the new `_directional_numeric_gradients` don't reuse input `Variable`s.  The change is OK because:
  - It's necessary for `_directional_numeric_gradients` to be clean;
  - Before the PR, `func` to be checked should not use `x.grad`.  I don't come up with a case `func` uses `x.grad`.

**Comment by toslunar** [comment]
Jenkins, test this please.

**Pull request merged**